tests/modules/saml/src/Controller/ServiceProviderTest.php

```diff
@@ -185,6 +208,247 @@ public function testWithAuthIdWithoutEntity(): void
+                    'ForceAuthn' => true
+                ],
+                false
+            ],
+            'target=https://evil.com' => [
+                [
+                    'target' => 'https://evil.com'
+                ],
+                [],
+                true
+            ]
+        ];
+    }
+
+    /**
+     * Test Login without Session/ Not Authenticated
+     *
+     * @dataProvider loginNotAuthenticatedDataProvider
+     *
+     * @param   array  $queryParameters
+     * @param   array  $options
+     * @param   bool   $expectingException
+     *
+     * @return void
+     */
+    public function testLoginHandleNotAuthenticated(array $queryParameters, array $options, bool $expectingException = false): void
+    {
+        $request = Request::create(
+            '/sp/login/phpunit',
+            'GET',
+            $queryParameters
+        );
+
+        $info = ['AuthId' => 'phpunit'];
+        $config =
+            [
+                'saml:SP',
+                'entityID' => 'urn:x-simplesamlphp:example-sp',
+                'RequestInitiation' => true,
+                'RelayState' => self::RELAY_STATE,
+            ];
+
+        $spSource = new class ($info, $config) extends Source\SP {
+            // stub
+        };
+
+        $as = $this->getMockBuilder(Auth\Simple::class)
+                   ->disableOriginalConstructor()
+                   ->getMock();
+        $as->method('isAuthenticated')->willReturn(false);
+        $as->method('getAuthSource')->willReturn($spSource);
+
+        if ($expectingException) {
+            $this->expectException(\SimpleSAML\Error\Exception::class);
+            $this->expectExceptionMessage('URL not allowed: https://evil.com');
+            $this->serviceProvider->callHandleLogin($request, $as, $spSource, $this->httpUtils);
+        } else {
+            $as->expects($this->once())->method('login')->with($options);
+            $returnsTo = $this->serviceProvider
+                              ->callHandleLogin($request, $as, $spSource, $this->httpUtils);
+            $this->assertEquals($options['ReturnTo'], $returnsTo);
+        }
+    }
+
+    /**
+     * @return array
+     */
+    public static function loginAuthenticatedDataProvider(): array
+    {
+        return [
+            'no query params' => [
+                [], // If no query params are provided, it should default to the RelayState
+                [
+                    'ReturnTo' =>  self::RELAY_STATE,
+                ],
+                false
+            ],
+            'entityID=example.edu' => [
+                [
+                    'entityID' => 'example.edu',
+                ],
+                [
+                    'ReturnTo' =>  self::RELAY_STATE,
+                ],
+                false
+            ],
+            'entityID=example.edu&forceAuthn=false&isPassive=false' => [
+                [
+                    'entityID' => 'example.edu',
+                    'forceAuthn' => false,
```

Comment:
I'm wondering if these should all be strings for the query parameters values.  e.g `'forceAtuhn' => 'false'` since in a real scenarios everything in `$_GET` is a string.

Reply:
done